libs/audio-recording/recording.ts

```diff
@@ -0,0 +1,279 @@
+            }
+            console.warn( "pxt-codal-audio: Impossible code state! Emergency reset of internal timing loop!" )
+            _isSetup = false
+        })
+    }
+
+    function __emitEvent__( type: AudioEvent ): void {
+        control.raiseEvent(AUDIO_EVENT_ID, AUDIO_VALUE_OFFSET+type, EventCreationMode.CreateAndFire )
+    }
+
+    function __setMode__( mode: AudioRecordingMode ): void {
+        switch( mode ) {
+            case AudioRecordingMode.Stopped:
+                if( _moduleMode == AudioRecordingMode.Recording ) {
+                    _moduleMode = AudioRecordingMode.Stopped
+                    return __emitEvent__( AudioEvent.StoppedRecording )
+                }
+                
+                if( _moduleMode == AudioRecordingMode.Playing ) {
+                    _moduleMode = AudioRecordingMode.Stopped
+                    return __emitEvent__( AudioEvent.StoppedPlaying )
+                }
+
+                _moduleMode = AudioRecordingMode.Stopped
+                return
+            
+            case AudioRecordingMode.Playing:
+                if( _moduleMode !== AudioRecordingMode.Stopped ) {
+                    __setMode__( AudioRecordingMode.Stopped )
+                }
+                
+                _moduleMode = AudioRecordingMode.Playing
+                return __emitEvent__( AudioEvent.StartedPlaying )
+            
+            case AudioRecordingMode.Recording:
+                if (_moduleMode !== AudioRecordingMode.Stopped) {
+                    __setMode__(AudioRecordingMode.Stopped)
+                }
+
+                _moduleMode = AudioRecordingMode.Recording
+                return __emitEvent__( AudioEvent.StartedRecording )
+        }
+    }
+
+    /**
+     * Record an audio clip for a maximum of 3 seconds
+     * 
+     * @param sync If true, block until we run out of memory!
+     */
+    //% block="record audio for 3(s)"
+    //% weight=70
+    export function startRecording(): void {
+        __init__()
+        eraseRecording();
+        record();
+        __setMode__( AudioRecordingMode.Recording )
+    }
+
+
+
+    /**
+     * Play recorded audio
+     */
+    //% block="play recording"
+    //% weight=60
+    export function playAudio(): void {
+        __init__()
+        _playbackHead = 0
+        if( !isEmpty() ) {
+            __setMode__(AudioRecordingMode.Playing)
+            play();
+        }
+        return
+    }
+
+    export function stopRecording(stopEvent: StopEvent): void {
+        __init__()
+        __setMode__(AudioRecordingMode.Stopped)
+        _playbackHead = 0
+        stop();
+        return
+    }
+
+    export function eraseRecording(): void {
+        __init__()
+        __setMode__(AudioRecordingMode.Stopped)
+        _playbackHead = 0
+        _memoryFill = 0
+        erase();
+        return
+    }
+
+    /**
+     * Do something based on what the audio is doing
+     */
+    //% block="on audio %eventType"
```

Comment:
```suggestion
    //% block="on audio $eventType"
```
I believe `$` is preferred over `%`